File: src/fortran/heat_mpi_2d.f90

#define N           27
#define MAX_ITER    4000
#define TOL         1e-4
#define MAX_TEMP    100.0

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Creating a module for globalizing the allocatable arrays
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
module grid
    real, dimension(:,:), allocatable ::T_old,T_new
end module grid


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Creating a module for globalizing mpi related data
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
module my_mpi_vars
    use mpi

  ! variables to store virtual topology information
    ! Global rank and number of MPI processes
    integer ::  grank, gprocs
    integer ::  globalN, localN
    ! Local rank and number of MPI processes
    integer lrank, lprocs
    integer :: ndims = 2                                    ! Since we are decomposing grid in 2D
    integer, dimension(2) :: dims                           ! placeholder for dimensions
    integer, dimension(2) :: mycoords                       ! placeholder for Cartesian coordinates
    integer, dimension(2) :: period                         ! placeholder for periodicity information
    integer :: reorder = 0

    integer :: x_axis = 1, y_axis = 0                       ! convinience variables for referencing axes
    integer :: right, left, up, down                        ! placeholder for ranks of nearest neighbours

    integer :: cartcomm                                     ! Declaration of new communicator
    integer :: ierr                                         ! error handler for MPI calls

    integer, dimension(8)  :: req                           ! placeholder for MPI Isend and Irecv requests
    integer, dimension(MPI_STATUS_SIZE,8)  :: status        ! placeholder for maintaining status of the MPI requests. Size defined in mpif.h
end module my_mpi_vars



!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
program heat_mpi_2d
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    use grid
    use mpi
    use my_mpi_vars

    implicit none


    ! Local rows and columns along with ghost region
    integer rows, cols;


    integer :: i,j,iter,rem
    real :: dT,dT_local



    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD,grank,ierr)
    call MPI_Comm_size(MPI_COMM_WORLD,gprocs,ierr)

    ! Create a Cartesian grid communicator
    call MPI_Dims_create(gprocs, ndims, dims, ierr)

    ! Let's create a new communicator with these dimensions
    period(1) = 0
    period(2) = 0
    call MPI_Cart_create (MPI_COMM_WORLD, ndims, dims, period, reorder,cartcomm, ierr)

    ! Know who my neighbors are
    call MPI_Comm_rank(cartcomm, lrank,ierr)                     ! Who am I in new communicator
    call MPI_Cart_coords(cartcomm, lrank, ndims, mycoords,ierr)  !Get my Cartesian coordinates
    call MPI_Comm_size(cartcomm, lprocs,ierr)

    call MPI_Cart_shift(cartcomm, x_axis, 1, left, right,ierr);
    call MPI_Cart_shift(cartcomm, y_axis, 1, up, down,ierr);

    ! decide my local N
    globalN = N
    localN =  globalN / dims(1)
    rem = mod(globalN,gprocs)
    if (grank == gprocs - 1) then
        localN = localN + rem
    endif
    rows = localN + 2
    cols = localN + 2

    ! allocate memory
    allocate(T_old(rows,cols))
    allocate(T_new(rows,cols))

    call initialize(rows,cols)
    iter = 0
    dT = MAX_TEMP

    do while ( (dT > TOL) .AND. (iter <= MAX_ITER) )
        ! update boundaries
        call halo_update(rows,cols)
        do i = 2,rows-1
            do j = 2,cols-1
               T_new(i,j) = 0.25 * ( T_old(i-1,j) + T_old(i+1,j) + T_old(i,j-1) + T_old(i,j+1) )
            end do
        end do

        ! Find the maximum error on each rank and let every rank know what is the maximum on any rank
        dT = 0.0
        dT_local = 0.0
        do i = 2,rows-1
            do j = 2,cols-1
               dT_local = max(abs(T_new(i,j) - T_old(i,j)) , dT)
               T_old(i,j) = T_new(i,j)
            end do
        end do
        call MPI_Allreduce(dT_local, dT, 1, MPI_REAL, MPI_MAX, cartcomm,ierr);
        iter=iter+1
    end do

    !writing the output to the file is optional and is there for validation.
    call write_grid(rows,cols,iter)
    call MPI_Barrier(cartcomm,ierr)
    !call print_grid(rows,cols,iter)

    if (lrank == 0) then
        if ((iter - 1) == MAX_ITER) then
            print *, "Reached maximum iterations ",iter,". Error = ", dT
        else
            print *, lrank,": Converged in ",iter," iterations with and error of ",dT
        end if
    end if

call MPI_FINALIZE(ierr)
end program heat_mpi_2d


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Initialize the grids with appropirate boundary and initial conditions
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

subroutine initialize(rows,cols)
    use grid
    use my_mpi_vars
    implicit none
    integer :: rows,cols
    integer :: i,j

    do i=1,rows
        do j=1,cols
                T_old(i,j) = 0.75 * MAX_TEMP
                T_new(i,j) = 0.75 * MAX_TEMP
        end do
    end do

    if (mycoords(2) == 0) then
        do i=1,rows
            T_old(i,1) = MAX_TEMP
            T_new(i,1) = MAX_TEMP
        end do
    end if
    if (mycoords(1) == dims(1)-1) then
        do j=1,cols
            T_old(rows,j) = MAX_TEMP
            T_new(rows,j) = MAX_TEMP
        end do
    end if
end subroutine initialize



!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Subroutine for Halo exchange
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
subroutine halo_update(rows,cols)
    use grid
    use my_mpi_vars
    implicit none
    integer :: rows,cols
    integer :: i,j
    integer :: tag =100         ! not important for us but MPI expects a value.

    ! We will need to make send and receive buffers to exchange left and right columns
    real, dimension(rows) :: sendbuf_l
    real, dimension(rows) :: sendbuf_r
    real, dimension(cols) :: sendbuf_u
    real, dimension(cols) :: sendbuf_d

    real, dimension(rows) :: recvbuf_l
    real, dimension(rows) :: recvbuf_r
    real, dimension(cols) :: recvbuf_u
    real, dimension(cols) :: recvbuf_d


    do i =1,rows
        sendbuf_l(i) = T_old(i,2)
        sendbuf_r(i) = T_old(i,cols-1)
        ! The recvbuf value will change only for those messages with destination rank > -1
        recvbuf_l(i) = T_old(i,1)
        recvbuf_r(i) = T_old(i,cols)
    end do

    do j = 1,cols
        sendbuf_u(j) = T_old(2,j)
        sendbuf_d(j) = T_old(rows-1,j)
        ! The recvbuf value will change only for those messages with destination rank > -1
        recvbuf_u(j) = T_old(1,j)
        recvbuf_d(j) = T_old(rows,j)
    end do

    call MPI_Irecv(recvbuf_u, cols, MPI_REAL, up, tag, cartcomm, req(5),ierr);
    call MPI_Irecv(recvbuf_d, cols, MPI_REAL, down, tag, cartcomm, req(6),ierr);
    call MPI_Irecv(recvbuf_l, rows, MPI_REAL, left, tag, cartcomm, req(7),ierr);
    call MPI_Irecv(recvbuf_r, rows, MPI_REAL, right, tag, cartcomm, req(8),ierr);

    call MPI_Isend(sendbuf_u, cols, MPI_REAL, up, tag, cartcomm, req(1),ierr)
    call MPI_Isend(sendbuf_d, cols, MPI_REAL, down, tag, cartcomm,req(2),ierr);
    call MPI_Isend(sendbuf_l, rows, MPI_REAL, left, tag, cartcomm, req(3),ierr);
    call MPI_Isend(sendbuf_r, rows, MPI_REAL, right, tag, cartcomm,req(4),ierr);

    call MPI_Waitall(8, req, status,ierr);

    !Copy the received columns from left and right neighbors into the ghost region
    do i=1,rows
        T_old(i,1) = recvbuf_l(i)
        T_old(i,cols) = recvbuf_r(i)
    end do

    do j=1,cols
        T_old(1,j) = recvbuf_u(j)
        T_old(rows,j) = recvbuf_d(j)
    end do
 end subroutine halo_update



!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! write grid
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
subroutine write_grid(rows,cols,iter)
    use grid
    use my_mpi_vars
    integer :: i,j,iter,rows,cols,root
    integer :: x,y,indx                            ! some scratch variables
    integer :: numelems_local,recv_buf_size
    ! following allocations are significant on to root rank
    integer :: recv_nelems(lprocs)                 ! number of elements in a vector to receive from each rank
    integer :: disp(lprocs)                        ! array to hold offsets where data from each rank will go in the receive buffer

    real, dimension(:), allocatable :: sendbuf,recvbuf      ! send buffer to send the data to root and recvbuffer is only signifcant to root
    character(len=256)   ::  fname                         ! output filename

    root=0
    numelems_local = localN * localN
    ! tell Writer how many elements each rank is going to send
    call MPI_Gather(numelems_local, 1, MPI_INT, recv_nelems, 1, MPI_INT, root,cartcomm,ierr)

    ! disp is an array of offsets or displacements
    ! i.e. start index where to receive vector from each rank (useful only on master)
    if (lrank == root) then
        disp(1) = 0
        do i = 2,lprocs
            disp(i) = disp(i-1)+ recv_nelems(i)
        end do
    end if

    ! Now prepare for gathering the actual data
    allocate(sendbuf(numelems_local))
    indx=1
    do i = 2, rows-1
       do j = 2, cols-1
            sendbuf(indx) = T_old(i,j)
            indx=indx+1
       end do
    end do
    recv_buf_size=0
    do i = 1,lprocs
        recv_buf_size = recv_buf_size + recv_nelems(i)
    end do

    if (lrank == root) then
        allocate(recvbuf(recv_buf_size))
    end if

    call MPI_Gatherv(sendbuf, numelems_local, MPI_REAL, recvbuf, recv_nelems,disp, MPI_REAL, root, cartcomm,ierr);

    if (lrank == root) then
        write(fname,'(A,I0,A)') 'grid_t',iter,'.txt'
        open (unit = 10, file = fname, status='replace')
        indx=0
        do y=1,recv_buf_size,dims(2)*numelems_local
            do i=1,numelems_local,localN
                do x=1,dims(1)*numelems_local,numelems_local
                    do j=1,localN
                        indx = (y-1) + (i-1) + (x-1) + j
                        write(10,fmt='((F10.5))',advance='no') recvbuf(indx)
                    end do
                end do
                write(10,*)
            end do
        end do
        close(10)
    end if
end subroutine write_grid






!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! debugging printer
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
subroutine print_grid(rows,cols,iter)
    use grid
    use my_mpi_vars
    implicit none
    integer :: rows,cols,iter
    integer :: i,j
    character(len=32) :: fname
    write(fname,'(A,I0,A,I0,A)') 'grid_',lrank,'_t',iter,'.txt'

    open (unit = 10, file = fname, status='replace')
    do i=2,rows-1
       write(10,*) (T_old(i,j), j=2,cols-1)
    end do
    close(10)
 end subroutine print_grid

File: src/c/heat_mpi_2d.c
/* 
 * 1D decomposition of the domain in the column dimension
 * such that each MPI task will get an equal number of rows.
 */

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define N			27
#define MAX_ITER	4000
#define TOL			1e-4
#define MAX_TEMP	100.0

float** allocate(float**);
void initialize(float**);
void halo_update(float**);
void write_grid(int, float**);
void print_grid(int, float**);


// Global rank and number of MPI processes
int grank, gprocs;
int globalN, localN;
// Local rank and number of MPI processes
int lrank, lprocs;
// Local rows and columns along with ghost region
int rows, cols;

// variables to store virtual topology information
int ndims = 2;			// Since we are decomposing grid in 2D
int dims[2] = { 0, 0 };  	// placeholder for dimensions
int mycoords[2] = { 0, 0 };	// placeholder for Cartesian coordinates
int period[2] = { 0, 0 };	// placeholder for periodicity information
int reorder = 0;

int x_axis = 1, y_axis = 0;
int right, left, up, down;	//placeholder for ranks of nearest neighbours

MPI_Comm cartcomm;		// Declaration of new communicator

int main(int argc, char *argv[]) {

	int i, j;
	float **T_old, **T_new;

	MPI_Init(&argc, &argv);
	MPI_Comm_rank(MPI_COMM_WORLD, &grank);
	MPI_Comm_size(MPI_COMM_WORLD, &gprocs);

	// Create a Cartesian grid communicator

	MPI_Dims_create(gprocs, ndims, dims);	// Get a good guess on dimensions

	// Let's create a new communicator with these dimensions
	MPI_Cart_create(MPI_COMM_WORLD, ndims, dims, period, reorder, &cartcomm);
	// Know who my neighbors are
	MPI_Comm_rank(cartcomm, &lrank);			// Who am I in new communicator
	MPI_Cart_coords(cartcomm, lrank, ndims, mycoords);//Get my Cartesian coordinates
	MPI_Comm_size(cartcomm, &lprocs);

	MPI_Cart_shift(cartcomm, x_axis, 1, &left, &right);
	MPI_Cart_shift(cartcomm, y_axis, 1, &up, &down);
	// decide my local N
	globalN = N;
	localN = (int) (globalN / dims[0]);
	int rem = globalN % gprocs;
	if (grank == gprocs - 1) {
		localN += rem;
	}
	rows = localN + 2;
	cols = localN + 2;

	T_old = allocate(T_old);
	T_new = allocate(T_new);

	initialize(T_old);
	initialize(T_new);
	int iter = 0;
	float dT = MAX_TEMP, dT_local = MAX_TEMP;

	while (dT > TOL && iter <= MAX_ITER) {

		halo_update(T_old);
		for (i = 1; i < rows - 1; i++) {
			for (j = 1; j < cols - 1; j++) {
				T_new[i][j] = 0.25
						* (T_old[i - 1][j] + T_old[i + 1][j] + T_old[i][j - 1]
								+ T_old[i][j + 1]);
			}
		}
		dT = 0.0;
		dT_local = 0.0;
		for (i = 1; i < rows - 1; i++) {
			for (j = 1; j < cols - 1; j++) {
				dT_local = fmaxf(fabsf(T_new[i][j] - T_old[i][j]), dT_local);
				T_old[i][j] = T_new[i][j];
			}
		}

		MPI_Allreduce(&dT_local, &dT, 1, MPI_FLOAT, MPI_MAX, cartcomm);
		iter++;
	}
	write_grid(iter-1,T_old);
	// Printing results
	if (lrank == 0) {
		if ((iter - 1) == MAX_ITER)
			printf("Reached maximum iterations %d. Error = %2.4f\n", iter - 1,
					dT);
		else
			printf("Converged in %d iterations with and error of %2.4f\n",
					iter - 1, dT);
	}

	MPI_Finalize();
	return 0;
}

float** allocate(float **T) {
	int i = 0, j = 0;

	// Allocate including some extra for ghost regions

	T = (float**) malloc(rows * sizeof(float*));
	if (T != NULL) {
		T[0] = (float*) malloc(rows * cols * sizeof(float));
	}
	if (T[0] != NULL) {
		for (i = 0; i < rows; i++)
			T[i] = (*T + i * cols);
	}
	return T;
}

void initialize(float **T) {
	int i, j;

	for (i = 0; i < rows; i++)
		for (j = 0; j < cols; j++)
			T[i][j] = 0.75 * MAX_TEMP;

	if (mycoords[0] == dims[0] - 1) {
		for (j = 0; j < cols; j++)
			T[rows - 1][j] = MAX_TEMP;
	}
	if (mycoords[1] == 0) {
		for (i = 0; i < rows; i++)
			T[i][0] = MAX_TEMP;
	}

}

void halo_update(float **T) {

	int i, j;
	// We will need to make send and receive buffers to exchange left and right columns
	float *sendbuf_l = (float*) malloc(rows * sizeof(float));
	float *sendbuf_r = (float*) malloc(rows * sizeof(float));
	float *sendbuf_u = (float*) malloc(cols * sizeof(float));
	float *sendbuf_d = (float*) malloc(cols * sizeof(float));

	float *recvbuf_l = (float*) malloc(rows * sizeof(float));
	float *recvbuf_r = (float*) malloc(rows * sizeof(float));
	float *recvbuf_u = (float*) malloc(cols * sizeof(float));
	float *recvbuf_d = (float*) malloc(cols * sizeof(float));

	for (i = 0; i < rows; i++) {
		sendbuf_l[i] = T[i][1];
		sendbuf_r[i] = T[i][cols - 2];
		// The recvbuf value will change only for those messages with destination rank > -1
		recvbuf_l[i] = T[i][0];;
		recvbuf_r[i] = T[i][cols - 1];
	}
	for (j=0; j < cols; j++){
		sendbuf_u[j] = T[1][j];
		sendbuf_d[j] = T[rows-2][j];
		// The recvbuf value will change only for those messages with destination rank > -1
		recvbuf_u[j] = T[0][j];
		recvbuf_d[j] = T[rows-1][j];
	}

	MPI_Request req[8];
	MPI_Status status[8];
	int tag = 100; // not important for us but MPI expects a value.

	MPI_Irecv(recvbuf_u, cols, MPI_FLOAT, up, tag, cartcomm, &req[4]);
	MPI_Irecv(recvbuf_d, cols, MPI_FLOAT, down, tag, cartcomm, &req[5]);
	MPI_Irecv(recvbuf_l, rows, MPI_FLOAT, left, tag, cartcomm, &req[6]);
	MPI_Irecv(recvbuf_r, rows, MPI_FLOAT, right, tag, cartcomm, &req[7]);

	MPI_Isend(sendbuf_u, cols, MPI_FLOAT, up, tag, cartcomm, &req[0]);
	MPI_Isend(sendbuf_d, cols, MPI_FLOAT, down, tag, cartcomm, &req[1]);
	MPI_Isend(sendbuf_l, rows, MPI_FLOAT, left, tag, cartcomm, &req[2]);
	MPI_Isend(sendbuf_r, rows, MPI_FLOAT, right, tag, cartcomm, &req[3]);

	MPI_Waitall(8, req, status);

	// Copy the received columns from left and right neighbors into the ghost region
	for (i = 0; i < rows; i++) {
		T[i][0] = recvbuf_l[i];
		T[i][cols - 1] = recvbuf_r[i];
	}
	for (j=0; j < cols; j++){
		T[0][j] = recvbuf_u[j];
		T[rows-1][j] = recvbuf_d[j];
	}
	free(sendbuf_l);
	free(sendbuf_r);
	free(sendbuf_u);
	free(sendbuf_d);
	free(recvbuf_l);
	free(recvbuf_r);
	free(recvbuf_u);
	free(recvbuf_d);
}

void print_grid(int iter, float **T) {
	int i, j;
	char fname[256];
	sprintf(fname, "grid%d_%d.txt", lrank, iter);
	FILE *fw = fopen(fname, "w+");
	fprintf(fw, "grank: %d, lrank: %d\n", grank, lrank);
	fprintf(fw, "r,l,u,d: %d,%d,%d,%d\n", right, left, up, down);
	fprintf(fw, "dims: %d,%d\n", dims[0], dims[1]);
	fprintf(fw, "Coords: %d,%d\n", mycoords[0], mycoords[1]);
	fprintf(fw, "globalN: %d, localN: %d, rows: %d, cols: %d\n", globalN,
			localN, rows, cols);
	for (i = 0; i < rows; i++) {
		for (j = 0; j < cols; j++) {
			fprintf(fw, "%2.2lf ", T[i][j]);
		}
		fprintf(fw, "\n");
	}
	fclose(fw);
}

void print_grid_1d(int iter, float *T) {
	int i, j;
	char fname[256];
	sprintf(fname, "grid%d_%d.txt", lrank, iter);
	FILE *fw = fopen(fname, "w+");
	fprintf(fw, "grank: %d, lrank: %d\n", grank, lrank);
	fprintf(fw, "r,l,u,d: %d,%d,%d,%d\n", right, left, up, down);
	fprintf(fw, "dims: %d,%d\n", dims[0], dims[1]);
	fprintf(fw, "Coords: %d,%d\n", mycoords[0], mycoords[1]);
	fprintf(fw, "globalN: %d, localN: %d, rows: %d, cols: %d\n", globalN,
			localN, rows, cols);
	int x=0;
	for (i = 1; i < rows - 1; i++){
		for (j = 1; j < cols - 1; j++){
			fprintf(fw, "%2.4lf ", T[x]);
			x++;
		}
		fprintf(fw, "\n");
	}
	fclose(fw);
}
void write_grid(int iter, float **T) {
	int i=0, j=0,indx=0;

	int root = 0;
	//Gather information about decomposition from all process onto Root process
	int total_elements = localN * localN;
	int* recv_numelems;
	int* disp;
	if (lrank == root) {
		recv_numelems = (int*) malloc(lprocs * sizeof(int));
		disp = (int*) malloc(lprocs * sizeof(int));
	}
	// tell Writer how many elements each rank is going to send
	MPI_Gather(&total_elements, 1, MPI_INT, recv_numelems, 1, MPI_INT, root,cartcomm);

	// disp is an array of offsets or displacements
	// i.e. start index where to receive vector from each rank (useful only on master)
	if (lrank == root) {
		disp[0] = 0;
		for (i = 1; i < lprocs; i++){
			disp[i] = disp[i-1]+recv_numelems[i];
		}
	}
	// Now prepare for gathering the actual data
	float *sendbuf = (float*) malloc(total_elements * sizeof(float));
	if (sendbuf == NULL){
		MPI_Abort(MPI_COMM_WORLD,911);
	}

	for (i = 1; i < rows - 1; i++){
		for (j = 1; j < cols - 1; j++){
			sendbuf[indx] = T[i][j];
			indx++;
		}
    }
    float* recv_buffer;
    if (lrank == root){
    	recv_buffer = (float*) malloc(globalN*globalN * sizeof(float));
    	if (recv_buffer == NULL)
    		MPI_Abort(MPI_COMM_WORLD,911);
    }
	MPI_Gatherv(sendbuf, total_elements, MPI_FLOAT, recv_buffer, recv_numelems,disp, MPI_FLOAT, root, cartcomm);

	if (lrank == root) {
		char fname[256];
		sprintf(fname, "grid_t%d.txt", iter);
		FILE *fw = fopen(fname, "w+");
		indx=0;

		int x,y;
		indx=0;
		for (y=0; y < (globalN*globalN); y+=(dims[1]*localN*localN)){
				for(i=0; i<localN*localN; i+=localN){
					for(x =0; x < dims[0]*localN*localN; x+=localN*localN){
						for(j=0;j<localN;j++){
							indx=y+i+x+j;
							fprintf(fw,"%2.5f ",recv_buffer[indx]);
						}
					}
					fprintf(fw,"\n");
				}
			}
		fclose(fw);

	}

}
